### ISSUE TITLE
TensorrtExecutionProvider slower than CUDAExecutionProvider: Transformers

### ISSUE THREAD

**Issue by oborchers** [microsoft/onnxruntime#7230]
**Describe the bug**
The CUDAExecutionProvider provides a 2x speedup over the TensorrtExecutionProvider, which is completely counterintuitive.

**Urgency**
None

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- ONNX Runtime installed from (source or binary): Source
- ONNX Runtime version: 1.7.1
- Python version: 3. 8
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: 11.1.1 / 8.0.5
- GPU model and memory: Nvidia V100

**To Reproduce**

1) Build the following Dockerfile (may take some time)

```
# Current version: v0.0.1

#FROM nvidia/cuda:10.2-cudnn8-devel-ubuntu18.04
FROM nvcr.io/nvidia/tensorrt:20.12-py3

# TensorRT: 7.2.2 (required for ONNX To TensorRT)
# Ubuntu 20.04
# Note: Container image 20.12-py3 contains Python 3.8.
# NVIDIA CUDA 11.1.1 including cuBLAS 11.3.0.
# NVIDIA cuDNN 8.0.5
# NVIDIA NCCL 2.8.3 (optimized for NVLink™ )
# Note: Although NCCL is packaged in the container, it does not affect TensorRT nor inferencing in any way.
# MLNX_OFED 5.1
# OpenMPI 4.0.5
# Nsight Compute 2020.2.1.8
# Nsight Systems 2020.3.4.32
# Source: https://docs.nvidia.com/deeplearning/tensorrt/container-release-notes/rel_20-12.html#rel_20-12

ENV CUDA_HOME=/usr/local/cuda
ENV CUDACXX=/usr/local/cuda/bin/nvcc
ENV CUDNN_HOME=/usr/local/cuda

##################################
# Basics                         #
##################################

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && \
    apt-get install -y --no-install-recommends \
    build-essential \
    git \
    libncurses5-dev \
    libgdbm-dev \
    libnss3-dev \
    libssl-dev \
    libreadline-dev \
    libffi-dev \
    libsqlite3-dev \
    libbz2-dev \
    liblzma-dev \
    zlib1g-dev \
    wget && \
    rm -rf /var/lib/apt/lists/*


##################################
# GCC                            #
##################################

# Only env vars are required for this image
ENV CC=/usr/bin/gcc
ENV CXX=/usr/bin/g++


###################################
# CMake                           #
###################################

ARG CMAKE_VERS=3.19.1
RUN apt-get update && \
    mkdir -p build && cd build && \
    wget https://github.com/Kitware/CMake/releases/download/v$CMAKE_VERS/cmake-$CMAKE_VERS.tar.gz && \
    tar -xvzf cmake-$CMAKE_VERS.tar.gz && \
    cd cmake-$CMAKE_VERS && \
    ./bootstrap && \
    make && \
    make install

ENV PATH /usr/local/nvidia/bin:/usr/local/cuda/bin:/code/cmake-$CMAKE_VERS-Linux-x86_64/bin:${PATH}


#####################################
# Minor additions and packages      #
#####################################

RUN apt-get update && \
    pip install -U pip && \
    pip install -U setuptools

RUN apt-get update && \
    apt-get install -y language-pack-en && \
    locale-gen en_US.UTF-8 && \
    update-locale LANG=en_US.UTF-8

RUN apt-get update && \
    apt-get install -y protobuf-compiler libprotoc-dev && \
    pip install -U pytest pytest-cov protobuf numpy==1.19.5

#####################################
# Compile ONNX Runtime and Install  #
#####################################

ENV ONNX_ML=1
RUN pip install -U onnx

ARG ONNXRUNTIME_VERS=1.7.1
RUN apt-get update && \
    git clone https://github.com/microsoft/onnxruntime.git --branch v$ONNXRUNTIME_VERS --single-branch && \
    /bin/sh onnxruntime/dockerfiles/scripts/install_common_deps.sh && \
    cd onnxruntime/ && \
    pip install -r requirements-dev.txt && \
    /bin/sh ./build.sh \
    --config Release \
    --update \
    --build \
    --build_wheel \
    --parallel \
    --use_cuda \
    --cuda_home $CUDA_HOME \
    --cudnn_home /usr/lib/x86_64-linux-gnu/ \
    --use_tensorrt \
    --tensorrt_home /opt/tensorrt/ \
    --build_dir /workspace/onnxruntime/python

RUN pip install onnxruntime-tools git+https://github.com/microsoft/ort-customops.git

RUN pip uninstall -y onnxruntime && \
    pip install -U /workspace/onnxruntime/python/Release/dist/*gpu*.whl

RUN echo "ONNXRuntime Execution Providers: " && python -c "import onnxruntime as ort; print(ort.get_available_providers())"

WORKDIR /home

```
2. Export the environment variables for tensorrt:
```
export ORT_TENSORRT_MAX_BATCH_SIZE=10
export ORT_TENSORRT_MAX_WORKSPACE_SIZE=4294967296
export ORT_TENSORRT_MAX_PARTITION_ITERATIONS=20
export ORT_TENSORRT_MIN_SUBGRAPH_SIZE=5
```
3. Check that all providers are available:
```
echo "ONNX Runtime Execution Providers: " && python -c "import onnxruntime as ort; print(ort.get_available_providers())"
> ['TensorrtExecutionProvider', 'CUDAExecutionProvider', 'CPUExecutionProvider']
```

4. Then, run the following python code:

```
import torch
import json
import numpy as np
import onnx
import onnx_tensorrt.backend as backend
import onnxruntime as rt
import multiprocessing
import transformers

from termcolor import colored
from transformers import convert_graph_to_onnx
from pathlib import Path
from sentence_transformers import SentenceTransformer
from onnxruntime_customops import get_library_path

print(rt.__version__)
print(transformers.__version__)

span = "Hello my friends!"

ONNX_PROVIDERS = ["CUDAExecutionProvider", "CPUExecutionProvider",]

opt = rt.SessionOptions()
opt.register_custom_ops_library(get_library_path())
opt.graph_optimization_level = rt.GraphOptimizationLevel.ORT_ENABLE_EXTENDED
opt.log_severity_level = 4
opt.intra_op_num_threads = multiprocessing.cpu_count()
opt.execution_mode = rt.ExecutionMode.ORT_SEQUENTIAL

model_name = "sentence-transformers/bert-base-nli-stsb-mean-tokens"
pipeline_name = "feature-extraction"
model_pth = Path(f"encoder/bert-base-nli-stsb-mean-tokens.onnx")

nlp = transformers.pipeline(pipeline_name, model=model_name, tokenizer=model_name, device=0)
tokenizer = nlp.tokenizer

if not model_pth.exists():
    convert_graph_to_onnx.convert(
        framework="pt",
        model=model_name,
        output=model_pth,
        opset=12,
        tokenizer=model_name,
        use_external_format= False,
        pipeline_name= pipeline_name,
    )
```

4. Time the TensorRT execution time in a notebook:
```
sess = rt.InferenceSession(str(model_pth), opt, providers=["TensorrtExecutionProvider"])
%%timeit
model_input = tokenizer.encode_plus(span)
model_input = {name : np.atleast_2d(value) for name, value in model_input.items()}
onnx_result = sess.run(None, model_input)
```
> 3.53 ms ± 145 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

4. Time the CUDA execution time in a notebook:
```
sess = rt.InferenceSession(str(model_pth), opt, providers=["CUDAExecutionProvider"])
%%timeit
model_input = tokenizer.encode_plus(span)
model_input = {name : np.atleast_2d(value) for name, value in model_input.items()}
onnx_result = sess.run(None, model_input)
```
> 2.2 ms ± 51.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

**Expected behavior**
At least to my intuition, TensorRT should be faster.

**Comment by pranavsharma** [comment]
@stevenlix can you please take a look? Thx.

**Comment by stevenlix** [comment]
TensorRT usually takes much longer time to build engine. Please add a warmup session (i.e. call sess.run once) before inference, so that engine can be built in advance. 
Another thing is if the model can't run as a whole in TensorRT, you may add CUDA as a fallback, for example, sess = rt.InferenceSession(str(model_pth), opt, providers=["TensorrtExecutionProvider", "CUDAExecutionProvider"])

**Comment by oborchers** [comment]
@stevenlix : Thank you for your comments! I have gone through all of those possibilities and the results are exactly the same. Furthermore, I provided a notebook for exact reproduction of the behavior here:
https://github.com/oborchers/Medium_Repo/blob/master/onnxruntime-issues/TensorRT%20Slow.ipynb
Dockerfile can be found here: https://github.com/oborchers/Medium_Repo/blob/master/onnxruntime-issues/Dockerfile

The speeds look as follows:

- Torch: 95 sentences/s
- CUDA + CPU: 384 sentences/s
- CUDA: 370 sentences/s
- Tensorrt + CUDA + CPU: 91 sentences/s
- Tensorrt: 91 sentences/s

Additional Info:
- Tensorrt runs have been performed with 50 warmup iterations
- GPU usage via `nvidia-smi` constantly switches between 0 and to around 30-40% maxing at 70% for Tensorrt. For CUDA+CPU, it is running at a constant 65 - 75%.

This is also something I have experience half a year ago while trying this, but postponed investigation as I assumed I just did something wrong. But now that I am able replicate the exact same problem in Docker, I can safely discard a problem on the side on our master server configuration.

GPUs otherwise not occupied.

**Comment by stevenlix** [comment]
If sequence length of the model is dynamic, you may need to build an engine in warmup to cover the shape range, otherwise TensorRT will rebuild engine during inference every time when new shape inputs are coming. For example, if sequence length varies from A to B, in warmup you may call session.run once with input shape[1,A] and call another session.run with input shape[1, B], then the engine will be used for any sequence length between A and B, and will not rebuilt during inference.

**Comment by oborchers** [comment]
@stevenlix:

This partially addresses the issue. I tried to do exactly that, and this has the side effect that the GPU usage now remains constantly at 85-90%. Encoding 10,000 sentences took 38s at 263 sentences/s. However, I am still nowhere near to beating base CUDA, which is residing at around 400-450 sentences/s.

Next I tried to pad all inputs to the maximum length and sub-lengths. The following padding sizes result in the following performance:

- 128: 90% GPU usage @ 192 sentences/s
- 256: 92% GPU usage @  128 sentences/s
- 512: 95% GPU usage @ 75 sentences/s

I also played around with the environment variables, but this does not seem to have any effect:
```
export ORT_TENSORRT_MAX_BATCH_SIZE=1
export ORT_TENSORRT_MAX_WORKSPACE_SIZE=8_589_934_592
export ORT_TENSORRT_MAX_PARTITION_ITERATIONS=40
export ORT_TENSORRT_MIN_SUBGRAPH_SIZE=1
```

**Comment by mudong0419** [comment]
I have the same problem. When inference with Bert(HuggingFace), TensorrtExecutionProvider is slower than CUDAExecutionProvider.

**Comment by LeeJiangWei** [comment]
> I have the same problem. When inference with Bert(HuggingFace), TensorrtExecutionProvider is slower than CUDAExecutionProvider.

Same here using HugginceFace models. It's a warm-up issue as menthioned in #8315, and it will be much faster the second time you call sess.run().
